processing/src/main/java/io/druid/segment/DimensionSelector.java

```diff
@@ -42,7 +44,12 @@
    *
    * Value cardinality would be 2.
    *
-   * @return the value cardinality
+   * Cardinality may be unknown (e.g. the selector used by IncrementalIndex while reading input rows),
+   * in which case this method will return -1. If cardinality is unknown, you should assume this
+   * dimension selector has no dictionary, and avoid calling "lookupId" or "lookupName" outside of
```

Comment:
This is a dangerous change. There are lots of places that will break horribly if (prior illegal value) -1 is returned.

I would feel more comfortable making an exception that is

```java
class UnavailableCardinalityException extends UOE{
}
```

and have things which can handle missing cardinality catch the error.